src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/VirtualDispatchHelpers.cs

```diff
@@ -0,0 +1,95 @@
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.InteropServices;
+
+namespace System.Runtime.CompilerServices;
+
+[StackTraceHidden]
+[DebuggerStepThrough]
+internal static unsafe partial class VirtualDispatchHelpers
+{
+    private struct VirtualResolutionData : IEquatable<VirtualResolutionData>
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public VirtualResolutionData(MethodTable* methodTable, IntPtr classHandleTargetMethod, IntPtr methodHandle)
+        {
+            HashCode = (int) ((uint)methodTable + (BitOperations.RotateLeft((uint)classHandleTargetMethod, 5)) + (BitOperations.RotateRight((uint)methodHandle, 5)));
+            MethodTable = methodTable;
+            ClassHandleTargetMethod = classHandleTargetMethod;
+            MethodHandle = methodHandle;
+        }
+        public int HashCode;
+        public MethodTable* MethodTable;
+        public IntPtr ClassHandleTargetMethod;
+        public IntPtr MethodHandle;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool Equals(VirtualResolutionData other) =>
+            HashCode == other.HashCode &&
+            MethodTable == other.MethodTable &&
+            ClassHandleTargetMethod == other.ClassHandleTargetMethod &&
+            MethodHandle == other.MethodHandle;
+
+        public override bool Equals(object? obj) => obj is VirtualResolutionData other && Equals(other);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public override int GetHashCode() => HashCode;
+    }
+
+    private struct VirtualFunctionPointerArgs
+    {
+        public IntPtr classHnd;
+        public IntPtr methodHnd;
+    };
+
+#if DEBUG
+        // use smaller numbers to hit resizing/preempting logic in debug
+        private const int InitialCacheSize = 8; // MUST BE A POWER OF TWO
+        private const int MaximumCacheSize = 512;
+#else
+        private const int InitialCacheSize = 128; // MUST BE A POWER OF TWO
+        private const int MaximumCacheSize = 8 * 1024;
+#endif // DEBUG
+
+    private static GenericCache<VirtualResolutionData, IntPtr> s_virtualFunctionPointerCache = new GenericCache<VirtualResolutionData, IntPtr>(InitialCacheSize, MaximumCacheSize);
+
+    [LibraryImport(RuntimeHelpers.QCall)]
+    private static unsafe partial IntPtr JIT_ResolveVirtualFunctionPointer(ObjectHandleOnStack obj, IntPtr classHandle, IntPtr methodHandle);
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    [DebuggerHidden]
+    private static unsafe IntPtr VirtualFunctionPointerSlowpath(object obj, IntPtr classHandle, IntPtr methodHandle)
+    {
+        IntPtr result = JIT_ResolveVirtualFunctionPointer(ObjectHandleOnStack.Create(ref obj), classHandle, methodHandle);
+        s_virtualFunctionPointerCache.TrySet(new VirtualResolutionData(RuntimeHelpers.GetMethodTable(obj), classHandle, methodHandle), result);
```

Comment:
NativeAOT follows this rule around not caching results (for interface dispatch, but appears to have the same problem for virtual function resolution)  CoreCLR does not appear to me to have logic in its dispatch behavior which avoids caching. Possibly we intended to build it, but I'm pretty sure we didn't. I can build a test case to validate my expectation if you'd like. The performance of the existing IDynamicInterfaceCastable implementation would likely be highly impacted if we change the rules here.

Reply:
The result of this is a useable model, although it's not well defined in behavior. The result is that if any type uses `IDynamicInterfaceCastable` ... the implementor of `IDynamicInterfaceCastable.GetInterfaceImplementaiton` really should provide an idempotent implementation for any case where it will successfully return an implementation, or they may have surprising behavior. However, we DO have the correct cache avoidance at cast time. This means that users of types which use `IDynamicInterfaceCastable` experience the expected behavior.

Reply:
> NativeAOT follows this rule around not caching results (for interface dispatch

We may want to fix NativeAOT to match CoreCLR.

Reply:
Agreed. I've opened https://github.com/dotnet/runtime/issues/107999 to track this issue